Use std::unique_ptr with custom deleter for spatialite and sqlite

m-kuhn opened the pull request:
We are not even close to where we should get.

* All those exec and error handling is still fragile.
* QgsSqliteHandle ~~does caching of connections. If we want to keep that, it's a bit more complicated to replace.~~ is happy to use the new RAII style

But hey, it's a start.

Related question, does anyone mind if we remove all those pre-spatialite-4 ifdefs? From what I can see, anything starting with trusty should be fine.

m-kuhn committed:
Use std::unique_ptr with custom deleter for spatialite and sqlite

nyalldawson commented on src/providers/spatialite/qgsspatialiteconnection.cpp:
```diff
@@ -65,11 +66,12 @@ QgsSpatiaLiteConnection::Error QgsSpatiaLiteConnection::fetchTables( bool loadGe
     return FailedToOpen;
 
-  int ret = checkHasMetadataTables( handle );
+  ret = checkHasMetadataTables( handle );
```
This looks a bit wrong - i think it should be database.get()

nyalldawson reviewed (commented):
Looking great!

m-kuhn committed:
Switch more places to using spatialite_database_unique_ptr

m-kuhn committed:
Use sq/spatialite unique_ptrs in QgsMapLayer and QgsOfflineEditing

m-kuhn committed:
Release spatialite context when a spatialite_database_unique_ptr scopes out

m-kuhn committed:
Switch new spatialite layer dialog to RAII spatialite

m-kuhn committed:
Switch map layer to RAII sqlite3

m-kuhn committed:
Switch QgsStyle to RAII sqlite3

m-kuhn committed:
Switch geopackage dataitem to RAII sqlite3

m-kuhn committed:
Const correctness for sqlite_database_unique_ptr

nyalldawson commented on src/app/qgsnewspatialitelayerdialog.cpp:
```diff
@@ -200,64 +200,58 @@ void QgsNewSpatialiteLayerDialog::pbnFindSRID_clicked()
+    while ( sqlite3_step( statement.get() ) == SQLITE_ROW )
     {
-      myCRSs.insert( QString::fromUtf8( ( const char * )sqlite3_column_text( ppStmt, 0 ) ) );
+      myCRSs.insert( QString::fromUtf8( ( const char * )sqlite3_column_text( statement.get(), 0 ) ) );
```
statement.columnAsText( 0 )

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -1357,44 +1357,39 @@ QString QgsMapLayer::saveNamedStyle( const QString &uri, bool &resultFlag )
 
     QString mySql = QStringLiteral( "create table if not exists tbl_styles(style varchar primary key,qml varchar)" );
-    myResult = sqlite3_prepare( myDatabase, mySql.toUtf8().data(), mySql.toUtf8().length(), &myPreparedStatement, &myTail );
+    database.prepare( mySql, myResult );
```
statement =

nyalldawson commented on src/core/qgsmaplayer.cpp:
```diff
@@ -1357,44 +1357,39 @@ QString QgsMapLayer::saveNamedStyle( const QString &uri, bool &resultFlag )
-
     mySql = QStringLiteral( "insert into tbl_styles(style,qml) values (?,?)" );
-    myResult = sqlite3_prepare( myDatabase, mySql.toUtf8().data(), mySql.toUtf8().length(), &myPreparedStatement, &myTail );
+    database.prepare( mySql, myResult );
```
statement =

nyalldawson commented on src/core/symbology/qgsstyle.cpp:
```diff
@@ -368,13 +367,15 @@ bool QgsStyle::load( const QString &filename )
     QDomDocument doc;
-    QString symbol_name = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( ppStmt, SymbolName ) ) );
-    QString xmlstring = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( ppStmt, SymbolXML ) ) );
+    QString symbol_name = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( statement.get(), SymbolName ) ) );
```
statement.columnAsText

nyalldawson commented on src/core/symbology/qgsstyle.cpp:
```diff
@@ -368,13 +367,15 @@ bool QgsStyle::load( const QString &filename )
-    QString symbol_name = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( ppStmt, SymbolName ) ) );
-    QString xmlstring = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( ppStmt, SymbolXML ) ) );
+    QString symbol_name = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( statement.get(), SymbolName ) ) );
+    QString xmlstring = QString::fromUtf8( reinterpret_cast< const char * >( sqlite3_column_text( statement.get(), SymbolXML ) ) );
```
statement.columnAsText

nyalldawson commented on src/core/symbology/qgsstyle.cpp:
```diff
@@ -1219,17 +1208,18 @@ QgsSymbolGroupMap QgsStyle::smartgroupsListMap()
   }
 
-  sqlite3_finalize( ppStmt );
+  statement.release();
```
Not really necessary, but no harm either

m-kuhn replied:
Actually, yes it does harm. Just removed all those, they'll prevent cleanup from happening (releasing just sets the internal pointer to `nullptr` without calling the deletor).

nyalldawson replied:
Oh man I wasted a couple of hours when I made that same mistake recently too - using .release() instead of .reset(). It's really tricky to pick up while reading over code.

IMO clang/gcc should throw a warning when .release() is called and nothing captures the returned value. There's no valid use case I can see for this situation.

nyalldawson commented:
really nice cleanups @m-kuhn !

m-kuhn commented:
Actually, stuck in the C++ hell right now.

Looks like member variables (`mSpatialiteContext`) are lost on returning unique_ptr derived types from a method. I was hoping for copy elision to happen but something goes wrong... The journey continues.

m-kuhn committed:
Move spatialite context management to deleter

m-kuhn committed:
Improve chances for copy elision

m-kuhn committed:
Prevent memory leaks

m-kuhn committed:
Use member functions over sqlite3 api

m-kuhn committed:
Avoid recreating spatialite handles

m-kuhn committed:
Code improvement

m-kuhn committed:
delete spatialite_database_unique_ptr::reset( sqlite3 * )

m-kuhn commented:
In it's current state I start to think that this can potentially be safe to use

m-kuhn committed:
Destroy spatialite context after closing database

According to instructions in https://groups.google.com/forum/#!msg/spatialite-users/83SOajOJ2JU/sgi5fuYAVVkJ
